File: 528-swapping-nodes-in-a-linked-list/swapping-nodes-in-a-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* swapNodes(ListNode* head, int k) {
        if (head == nullptr || head->next == nullptr){
            return head;
        }

        ListNode *curr = head, *first = head, *second = head;
        int a = 0;
        while (curr){
            a+=1;
            if (a == k){
                first = curr;
            }
            if (a >= k+1){
                second = second->next;
            }
            curr = curr->next;
        }

        int temp = first->val;
        first->val = second->val;
        second->val = temp;

        return head;
    }
};

File: 528-swapping-nodes-in-a-linked-list/swapping-nodes-in-a-linked-list.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func swapNodes(head *ListNode, k int) *ListNode {
    if head == nil || head.Next == nil { return head }

    first, second, curr := head, head, head
    cnt := 0

    for curr != nil {
        cnt++

        if cnt == k {
            first = curr
        } else if cnt > k {
            second = second.Next
        }

        curr = curr.Next
    }

    first.Val, second.Val = second.Val, first.Val

    return head
}